File: subprojects/tools/src/agent-tools.js
import fs from "node:fs/promises";
import { appendEvent, collapseLastOpenScope, createState, normalizeState, readState, writeState } from "#strap/core/state";
import { toolResult } from "#strap/core/result";
import { objectSchema, optionalString, requireString, stringSchema } from "#strap/core/schemas";

async function forkAgent(args) {
  const input = requireString(args, "state_path");
  const output = requireString(args, "child_state_path");
  const prompt = optionalString(args, "prompt", "");
  const parent = normalizeState(await readState(input));
  const child = structuredClone(parent);
  child.parent = { state_path: input, fork_prompt: prompt, created_at: new Date().toISOString() };
  appendEvent(child, { from: "harness", to: ["assistant"], kind: "fork", text: prompt });
  await writeState(output, child);
  return toolResult({ child_state_path: output });
}

async function foldAgent(args) {
  const parentPath = requireString(args, "state_path");
  const childPath = requireString(args, "child_state_path");
  const summary = requireString(args, "summary");
  const parent = normalizeState(await readState(parentPath));
  const child = normalizeState(await readState(childPath));
  appendEvent(parent, {
    from: "harness",
    to: ["assistant", "user"],
    kind: "agent_fold",
    text: summary,
    hidden: { child_state: child },
  });
  await writeState(parentPath, parent);
  return toolResult({ state_path: parentPath, folded_child_state_path: childPath });
}

async function contextPush(args) {
  const statePath = requireString(args, "state_path");
  const label = requireString(args, "label");
  const state = normalizeState(await readState(statePath));
  state.root.children.push({ type: "scope", label, status: "open", participants: ["assistant", "harness"], children: [] });
  await writeState(statePath, state);
  return toolResult({ state_path: statePath, label });
}

async function contextPop(args) {
  const statePath = requireString(args, "state_path");
  const summary = requireString(args, "summary");
  const state = normalizeState(await readState(statePath));
  collapseLastOpenScope(state, summary);
  await writeState(statePath, state);
  return toolResult({ state_path: statePath, summary });
}

async function initState(args) {
  const statePath = requireString(args, "state_path");
  await fs.writeFile(statePath, `${JSON.stringify(createState(), null, 2)}\n`);
  return toolResult({ state_path: statePath });
}

export function agentTools() {
  return [
    {
      name: "agent_fork",
      description: "Fork a child state file for tangential or delegated work. The child can later be folded back with a visible summary and hidden retained state.",
      inputSchema: objectSchema({
        state_path: stringSchema("Parent state JSON path"),
        child_state_path: stringSchema("Output child state JSON path"),
        prompt: stringSchema("Task or branch prompt for the child"),
      }, ["state_path", "child_state_path"]),
      readOnly: false,
      execute: forkAgent,
    },
    {
      name: "agent_fold",
      description: "Fold a child state back into its parent as a visible summary plus hidden retained child state.",
      inputSchema: objectSchema({
        state_path: stringSchema("Parent state JSON path to update"),
        child_state_path: stringSchema("Child state JSON path to fold"),
        summary: stringSchema("Visible summary of the child work"),
      }, ["state_path", "child_state_path", "summary"]),
      readOnly: false,
      execute: foldAgent,
    },
    {
      name: "context_push",
      description: "Open a compactable scope in a state file for exploratory work.",
      inputSchema: objectSchema({ state_path: stringSchema("State JSON path"), label: stringSchema("Scope label") }, ["state_path", "label"]),
      readOnly: false,
      execute: contextPush,
    },
    {
      name: "context_pop",
      description: "Collapse the most recent open scope into a visible summary while retaining the full subtree as hidden state.",
      inputSchema: objectSchema({ state_path: stringSchema("State JSON path"), summary: stringSchema("Visible scope summary") }, ["state_path", "summary"]),
      readOnly: false,
      execute: contextPop,
    },
    {
      name: "state_init",
      description: "Create an empty strap v0.2 actor/event/scope state file.",
      inputSchema: objectSchema({ state_path: stringSchema("State JSON path to create") }, ["state_path"]),
      readOnly: false,
      execute: initState,
    },
  ];
}

File: subprojects/cli/commands/run-calls/src/run_calls/agent_tools.clj
(ns run-calls.agent-tools
  (:require [run-calls.common :as c]
            [run-calls.state :as state]))

(defn fork-agent [args _]
  (let [parent (state/read-state (:state_path args))
        child (-> parent
                  (assoc :parent {:state_path (:state_path args) :fork_prompt (or (:prompt args) "") :created_at (str (java.time.Instant/now))})
                  (state/append-event {:from "harness" :to ["assistant"] :kind "fork" :text (or (:prompt args) "")}))]
    (state/write-state (:child_state_path args) child)
    (c/tool-result {:child_state_path (:child_state_path args)})))

(defn fold-agent [args _]
  (let [parent (state/read-state (:state_path args)) child (state/read-state (:child_state_path args))
        next (state/append-event parent {:from "harness" :to ["assistant" "user"] :kind "agent_fold" :text (str (:summary args)) :hidden {:child_state child}})]
    (state/write-state (:state_path args) next)
    (c/tool-result {:state_path (:state_path args) :folded_child_state_path (:child_state_path args)})))

(defn context-push [args _]
  (let [s (update-in (state/read-state (:state_path args)) [:root :children] conj {:type "scope" :label (str (:label args)) :status "open" :participants ["assistant" "harness"] :children []})]
    (state/write-state (:state_path args) s)
    (c/tool-result {:state_path (:state_path args) :label (:label args)})))

(defn context-pop [args _]
  (let [s (state/collapse-last-open-scope (state/read-state (:state_path args)) (str (:summary args)))]
    (state/write-state (:state_path args) s)
    (c/tool-result {:state_path (:state_path args) :summary (:summary args)})))

(defn init-state [args _]
  (spit (:state_path args) (str (c/json-str (state/create-state)) "\n"))
  (c/tool-result {:state_path (:state_path args)}))

(def tools {"agent_fork" fork-agent "agent_fold" fold-agent "context_push" context-push "context_pop" context-pop "state_init" init-state})
